### ISSUE TITLE
Disk I/O slow for bind mounted volumes in 1.12.1

### ISSUE THREAD

**Issue by tim-ireland** [docker/for-mac#673]
### Expected behavior
I upgraded from 1.12.0 to 1.12.1, and expected continued operation of development environment.

### Actual behavior
Disk I/O for bind mounted volumes became extremely slow, practically unusable.

### Information

  - Full output of the diagnostics from "Diagnose & Feedback" in the menu

```
Docker for Mac: version: 1.12.1 (2d5b4d9)
OS X: version 10.11.6 (build: 15G1004)
logs: /tmp/89AEC452-6BEA-4A86-B167-A82B55C4C1F1/20160919-133829.tar.gz
[OK]     docker-cli
[OK]     virtualization kern.hv_support
[OK]     menubar
[OK]     moby-syslog
[OK]     dns
[OK]     disk
[OK]     system
[OK]     app
[OK]     osxfs
[OK]     virtualization VT-X
[OK]     db
[OK]     slirp
[OK]     logs
[OK]     env
[OK]     vmnetd
[OK]     moby-console
[OK]     moby
[OK]     driver.amd64-linux
```

# docker info
```
Containers: 5
 Running: 0
 Paused: 0
 Stopped: 5
Images: 17
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 39
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.20-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: BZBS:KRVN:AKKG:NSON:JICX:SQHZ:Z5JW:SV2Z:2F2W:O7GJ:LLUV:Y7PB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 17
 Goroutines: 29
 System Time: 2016-09-19T17:39:19.593685869Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

### Steps to reproduce the behavior

  1. `docker run -it -v /tmp:/foo ubuntu`
  2. `time dd if=/dev/zero of=test.dat bs=1024 count=100000`

```
root@24c218753c4f:/# time dd if=/dev/zero of=test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
102400000 bytes (102 MB, 98 MiB) copied, 0.189994 s, 539 MB/s

real	0m0.194s
user	0m0.010s
sys	0m0.180s
```

  3. cd /foo
  4. time dd if=/dev/zero of=test.dat bs=1024 count=100000

```
root@24c218753c4f:/foo# time dd if=/dev/zero of=test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
102400000 bytes (102 MB, 98 MiB) copied, 21.4836 s, 4.8 MB/s

real	0m21.489s
user	0m0.100s
sys	0m1.570s
```

**Comment by tim-ireland** [comment]
I also tried:
* resetting to factor defaults
* uninstalling / reinstalling 1.12.1
* uninstalling 1.12.1 and reverting to 1.12.0

I expected the last step to at least help in the mean time - but after downgrading (happened to have old installing in my downloads folder) saw the same behavior in 1.12.0.  Is there something that might be left behind by the uninstaller?

Finally:
* Tested 1.12.1 on another mac as well - saw same performance issue.

**Comment by hanspagel** [comment]
same here

**Comment by genei09** [comment]
I can confirm this behavior as well.  Running simple operations like `git status` can take several seconds to complete

**Comment by theanimal666** [comment]
here to... my test suit runs now more then 2m instead of 20 seconds before the update was installed. Docker for mac 1.12.1

**Comment by lattwood** [comment]
We've narrowed down performance deltas being due to one of the two following factors

* NCQ support on the host volume device and/or
* PCIe SSD

We stat ~1000 files in a bind mount regularly in one of our containers and only one computer in the office has issues with doing that in < 5 seconds- it lacks NCQ and it's SSD is SATA based.

**Comment by tim-ireland** [comment]
@lattwood, thanks for the update.  The following is from the SATA device info on my system:

```
Apple SSD Controller:

  Vendor:	Apple
  Product:	SSD Controller
  Physical Interconnect:	PCI
  Link Width:	x4
  Link Speed:	8.0 GT/s
  Description:	AHCI Version 1.30 Supported

APPLE SSD SM0256G:

  Capacity:	251 GB (251,000,193,024 bytes)
  Model:	APPLE SSD SM0256G                       
  Revision:	BXW1SA0Q
  Serial Number:	S29CNYAG717137      
  Native Command Queuing:	Yes
  Queue Depth:	32
  Removable Media:	No
  Detachable Drive:	No
  BSD Name:	disk0
  Medium Type:	Solid State
  TRIM Support:	Yes
  Partition Map Type:	GPT (GUID Partition Table)
  S.M.A.R.T. status:	Verified
  Volumes:
EFI:
  Capacity:	209.7 MB (209,715,200 bytes)
  BSD Name:	disk0s1
  Content:	EFI
  Volume UUID:	0E239BC6-F960-3107-89CF-1C97F78BB46B
disk0s2:
  Capacity:	250.14 GB (250,140,434,432 bytes)
  BSD Name:	disk0s2
  Content:	Apple_CoreStorage
Recovery HD:
  Capacity:	650 MB (650,002,432 bytes)
  BSD Name:	disk0s3
  Content:	Apple_Boot
  Volume UUID:	9ED9D2A1-7434-3A63-B629-993A865FEC4A
```

**Comment by genei09** [comment]
my stats for posterity:
```
Apple SSD Controller:

  Vendor:	Apple
  Product:	SSD Controller
  Physical Interconnect:	PCI
  Link Width:	x4
  Link Speed:	8.0 GT/s
  Description:	AHCI Version 1.30 Supported

APPLE SSD SM0256G:

  Capacity:	251 GB (251,000,193,024 bytes)
  Model:	APPLE SSD SM0256G                       
  Revision:	BXW1SA0Q
  Serial Number:	<redacted>
  Native Command Queuing:	Yes
  Queue Depth:	32
  Removable Media:	No
  Detachable Drive:	No
  BSD Name:	disk0
  Medium Type:	Solid State
  TRIM Support:	Yes
  Partition Map Type:	GPT (GUID Partition Table)
  S.M.A.R.T. status:	Verified
  Volumes:
EFI:
  Capacity:	209.7 MB (209,715,200 bytes)
  BSD Name:	disk0s1
  Content:	EFI

disk0s2:
  Capacity:	250.14 GB (250,140,434,432 bytes)
  BSD Name:	disk0s2
  Content:	Apple_CoreStorage
Recovery HD:
  Capacity:	650 MB (650,002,432 bytes)
  BSD Name:	disk0s3
  Content:	Apple_Boot
```

**Comment by blackstar257** [comment]
Does anyone have any good work arounds?

**Comment by mingp** [comment]
We're also encountering a large (>10x) slowdown on a write-heavy disk IO workload, which might be this same issue. More specifically, we have some automated tests which reset MySQL database state afterwards by truncating all tables in the test database, and this process has recently gone from <5 seconds to ~1 minute. This happens on both the beta channel (Version 1.12.1-beta26.1 (build: 12100)) and the stable channel (Version 1.12.1 (build: 12133)).

Is there some way to revert to or install with an older version of Docker for Mac in the interim until this issue is resolved? Does anyone have links to older releases?

Thanks in advance.

**Comment by JanBednarik** [comment]
@mingp Here is link to download 1.12.0 stable: https://download.docker.com/mac/stable/1.12.0.11213/Docker.dmg

**Comment by dsheets** [comment]
It is likely that many of you are experiencing #668. We really recommend trying the solution in #668 and using Beta 27 if this is a blocking issue for you. 1.12.0-a has many known issues so reverting to it is not recommended. If you continue to experience poor performance, you may be experiencing #77 for which improvements are under way.

**Comment by mingp** [comment]
@dsheets 

Thank you for the info. I'll try that workaround and see what happens.

@JanBednarik 

Thank you for the link. I'll try the version rollback as a last resort, if the other solutions don't work, just to get our current development workflow unblocked for the short term.

**Comment by tim-ireland** [comment]
For the record, I tried the suggestion in #668, which was to use 1.12.2-rc1-beta27 (build: 12496) and `echo false > com.docker.driver.amd64-linux/disk/full-sync-on-flush` etc.

I'm still getting around 4 MB/s write time.

```
[root@a129def8e1b foo]# time dd if=/dev/zero of=test.dat bs=1024 count=100000

100000+0 records in
100000+0 records out
102400000 bytes (102 MB) copied, 23.8144 s, 4.3 MB/s

real	0m23.825s
user	0m0.070s
sys	0m1.400s
[root@a129def8e1b foo]#
```

**Comment by dsheets** [comment]
@tim-ireland could you check the performance of your original use case? Is it back to the old speed that you remember?

For the particular microbenchmark that you have used, we are aware that performance isn't where it needs to be but we also don't believe that there was a performance regression in the file system sharing component between 1.12.0 and 1.12.1. General performance improvement of file sharing is tracked by #77.

**Comment by tim-ireland** [comment]
@dsheets 

Thanks, I don't really have a good benchmark metric for comparison from before 1.12.1, but my test suite does seem to be running faster now with 1.12.2-rc1-beta27.   Definitely not as fast as it once was, but improved.  A portion of the test suite interacts with PostgreSQL, so I'm guessing that is where the performance gain is the highest.

**Comment by dsheets** [comment]
Great! Glad to hear it. The numbers you posted earlier are very similar to the numbers for the same microbenchmark on modern Mac's with SSDs that we've been seeing for a couple months. We have a number of improvements in the pipeline.

Given that this issue seems to be mostly covered by #668 and #77 and we know of no changes that would significantly negatively or positively impact shared file system performance between 1.12.0 and 1.12.1, I am going to close this issue and invite you to subscribe to #668 or #77 for more information about virtual block device and shared file system performance respectively.

Thanks @tim-ireland for your report and everyone else in the thread for your help. If you feel that there are additional performance regressions that haven't been adequately addressed, please feel free to re-open this issue, comment, or open a new issue for a different problem.

**Comment by mingp** [comment]
The workaround in #668 worked for me. Performance is now back to where it was before.

Thank you all for your help.
